Fix: `OtlpExporterOptions` is not picking up env variables

lyra95 opened the pull request:
Fixes #6558 

Changes

Update the constructor to augment the provided IConfiguration with environment variables before applying it.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

martincostello commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:
```diff
@@ -48,16 +48,10 @@ public class OtlpExporterOptions : IOtlpExporterOptions
-              configuration: new ConfigurationBuilder().AddEnvironmentVariables().Build(),
-              configurationType,
-              defaultBatchOptions: new())
+            configuration: new ConfigurationBuilder().Build(),
```
This is redundant - this now just creates an empty configuration. It could be refactored to avoid the need to do this.

lyra95 replied:
Changed to `IConfiguration?` and passed null here

martincostello commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs:
```diff
@@ -67,8 +61,12 @@ internal OtlpExporterOptions(
     {
         Debug.Assert(defaultBatchOptions != null, "defaultBatchOptions was null");
+        var finalConfiguration = new ConfigurationBuilder()
+            .AddEnvironmentVariables(OtlpSpecConfigDefinitions.EnvVarPrefix)
```
This _might_ have unintended consequences if only environment variables with this prefix are consumed now.

martincostello replied:
I _think_ it makes sense for any specified `IConfiguration` values to override the environment variables.

martincostello reviewed (commented):
The changes look reasonable.

Can you add a test that would fail without this change to protect against regressions in the future, and also update the CHANGELOG please?

lyra95 commented:
> The changes look reasonable.
> 
> Can you add a test that would fail without this change to protect against regressions in the future, and also update the CHANGELOG please?

I’ve added tests and updated the CHANGELOG. I just noticed that this change also fixes #5586, so I included a test case for it as well.

martincostello commented:
Maybe another way to achieve the same end result is to make the `IConfiguration` field nullable and do something like this:

```csharp

configuration ??= new ConfigurationBuilder().AddEnvironmentVariables().Build();
this.ApplyConfiguration(configuration, configurationType);
```

martincostello commented:
Does it not move the decision later in a way that is equivalent to before ("no `IConfiguration` was provided at all, so we create one with only environment variables in it")?

lyra95 commented:
The problematic case is "IConfiguration is provided, but no environment variables are in it".

Also, when [docs](https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/#otel_exporter_otlp_endpoint) says it is configurable via environment variables, I think the implementation should do so. (not "if environment variables are set and if `Host` is configured *appropriately*")

It would be ideal if `OtlpExporterOptions` reads directly from environment variables rather than from `IConfigrutaion`, but I think `AddEnvironmentVariables` to existing `IConfiguration` is an acceptable solution.
